lib/puppet/provider/foreman_resource/rest_v3.rb

```diff
@@ -95,6 +98,20 @@ def success?(response)
   end
 
   def error_message(response)
-    JSON.parse(response.body)['error']['full_messages'].join(' ') rescue "unknown error (response #{response.code})"
+    explanations = {
+      '400' => 'Something is wrong with the data sent to Foreman server',
+      '401' => 'Often this is caused by invalid Oauth credentials',
+      '404' => 'The requested resource was not found',
+      '500' => 'Check /var/log/foreman/production.log on Foreman server for detailed information',
+      '502' => 'The webserver received an invalid response from the backend service. Was Foreman unable to handle the request?',
+      '503' => 'The webserver was unable to reach the backend service. Is foreman.service running?',
+      '504' => 'The webserver timed out waiting for a response from the backend service. Is Foreman under unusually heavy load?'
+    }
+
+    if explanation = explanations[response.code.to_str]
```

Comment:
The convention is to use braces when there is an assignment as a visual cue. Otherwise it's too easy to confuse with `==` and it may just be a bug/typo.
```suggestion
    if (explanation = explanations[response.code.to_str])
```

The alternative is:
```suggestion
    explanation = explanations[response.code.to_str]
    if explanation
```

Reply:
Thanks, I wasn't aware of this convention.

Thinking about it some more, would this be acceptable?

```
if !(explanation = explanations[response.code.to_str]).nil?
```

The `()`s are only useful if the reader knows the convention.

Reply:
Nevermind, it's not even semantically equivalent that way...

Reply:
My plan is to add Rubocop to the repo which will enforce the convention. If you don't do it and meant `if x == y`, Rubocop will ask you to change it to `if (x = y)`. That should tell you enough.

On a side note, Python thought it was a bad idea to allows assignment statements in if conditions to prevent those logic bugs. Recently it did add `if x := y` as a compromise.

That's the theory behind it.